KAFKA-5192 WindowStore range scan KIP-155

xvrl opened the pull request:
Implements range scan for keys in windowed and session stores

Modifies caching session and windowed stores to use segmented cache keys.
Cache keys are internally prefixed with their segment id to ensure key ordering in the cache matches the ordering in the underlying store for keys spread across multiple segments.
This should also result in fewer cache keys getting scanned for queries spanning only some segments.

xvrl committed:
range scan implementation for window / session stores

xvrl committed:
fix Bytes.get() javadocs

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/ReadOnlySessionStore.java:
```diff
@@ -41,4 +41,6 @@
      * @return   KeyValueIterator containing all sessions for the provided key.
      */
     KeyValueIterator<Windowed<K>, AGG> fetch(final K key);
+
```
I guess we need javadoc for these new methods on the public interface

xvrl replied:
done

xvrl committed:
update to latest API

xvrl committed:
simplify optimized bytes iterator, fix key bytes

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractMergedSortedCacheStoreIterator.java:
```diff
@@ -29,27 +28,26 @@
 
     abstract K deserializeCacheKey(final Bytes cacheKey);
 
+    abstract V deserializeCacheValue(LRUCacheEntry cacheEntry);
```
nit: `final`

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -103,6 +103,25 @@ public void apply(final List<ThreadCache.DirtyEntry> entries) {
+                                                           long earliestSessionEndTime,
+                                                           long latestSessionStartTime) {
+        validateStoreOpen();
+        final Bytes binarySessionIdFrom = Bytes.wrap(keySerde.serializer().serialize(topic, keyFrom));
```
These should use `serdes.rawKey(...)` as has been done in: https://github.com/apache/kafka/pull/2963

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStore.java:
```diff
@@ -86,4 +86,47 @@ public void remove() {
     }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, V> fetch(K from, K to) {
```
This method is basically identical to `fetch(final K key)`.  Maybe create an interface, `Fetcher` (or similar) with 2 impls, one that delegates to `store.fetch(from, to)` and the other that delegates to `store.fetch(key)`. Then make both `fetch` methods delegate to another internal fetch method passing in the appropriate impl?

xvrl replied:
sounds reasonable

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStore.java:
```diff
@@ -86,4 +86,47 @@ public void remove() {
     }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, V> fetch(K from, K to) {
```
nit: `final`

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java:
```diff
@@ -84,4 +86,46 @@ public void remove() {
     }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, V> fetch(K from, K to, long timeFrom, long timeTo) {
```
nit: `final`

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java:
```diff
@@ -84,4 +86,46 @@ public void remove() {
+
+    @Override
+    public KeyValueIterator<Windowed<K>, V> fetch(K from, K to, long timeFrom, long timeTo) {
+        final List<ReadOnlyWindowStore<K, V>> stores = provider.stores(storeName, windowStoreType);
```
Same as in `CachingSessionStore`. Both `fetch` methods are basically identical.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheKeyValueStoreIterator.java:
```diff
@@ -46,6 +49,11 @@ K deserializeCacheKey(final Bytes cacheKey) {
     }
 
     @Override
+    V deserializeCacheValue(LRUCacheEntry cacheEntry) {
```
nit: `final`

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java:
```diff
@@ -0,0 +1,72 @@
+    private final StateSerdes<K, V> serdes;
+    private final long windowSize;
+
+    public MergedSortedCacheWindowStoreKeyValueIterator(
```
can be package private

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java:
```diff
@@ -0,0 +1,72 @@
+    private final long windowSize;
+
+    public MergedSortedCacheWindowStoreKeyValueIterator(
+        PeekingKeyValueIterator<Bytes, LRUCacheEntry> filteredCacheIterator,
```
nit: `final` here and elsewhere

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:
```diff
@@ -55,6 +55,18 @@
     }
 
     @Override
+    public KeyValueIterator<Bytes, byte[]> fetch(final Bytes keyFrom, Bytes keyTo, final long from, final long to) {
```
Could we make the other `fetch` method delegate to this one?

xvrl replied:
underlying byte ranges differ significantly for single keys vs key range, so I kept both methods

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java:
```diff
@@ -114,7 +125,18 @@
          * @return
          */
-        HasNextCondition hasNextCondition(final Bytes binaryKey, long from, long to);
+        HasNextCondition hasNextCondition(final Bytes binaryKey, final long from, final long to);
```
Could we remove this method and just have all the `fetch` calls use the other one?

xvrl replied:
my only concern with removing the single key method would be the potential performance impact of having to check both `<=` and `>=` for each value as opposed to just `=` for single key lookups.

dguy replied:
Ok - i guess i'd like to understand what the performance impact is.

xvrl replied:
I'm happy to use a single method for now. In the essence of time. If there is a clear need for both, I'll do a follow-up PR.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java:
```diff
@@ -72,6 +72,27 @@ public boolean hasNext(final KeyValueIterator<Bytes, ?> iterator) {
     }
 
     @Override
+    public HasNextCondition hasNextCondition(final Bytes binaryKeyFrom, final Bytes binaryKeyTo, final long from, final long to) {
```
As noted above - can we remove the other overload of this method and just use this one?

xvrl replied:
removed

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java:
```diff
@@ -68,6 +68,28 @@ public boolean hasNext(final KeyValueIterator<Bytes, ?> iterator) {
     }
 
     @Override
+    public HasNextCondition hasNextCondition(final Bytes binaryKeyFrom, final Bytes binaryKeyTo, final long from, final long to) {
```
Same as above

xvrl committed:
some refactoring and more tests

xvrl committed:
include window end in Windowed.toString()

xvrl committed:
checksytle

xvrl committed:
delegate to more general method

xvrl committed:
segment cache keys to maintain key ordering between cache and underlying store

xvrl committed:
fix window/session abitrary key ranges

xvrl commented:
@dguy addressed comments in your initial review, and also expanded test coverage.
After some testing I realized we needed to have the cache also respect the segmentation of the underlying store to have a consistent ordering between cache and underlying store, when querying for key ranges across segments.

This led me to change the caching stores to prefix every cache key with a segment id internally. One benefit of that, is that time range queries will now only scan relevant segments in the cache as well, potentially reducing the number of cache key to scan.

There were also some gotchas with respect to key ranges which were a bit tricker, so I added the upper/lowerRangeFixedSize to keep the optimized ranges for single key queries

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:
```diff
@@ -149,23 +154,54 @@ public synchronized void put(final K key, final V value, final long timestamp) {
     }
 
+    @Override
+    public KeyValueIterator<Windowed<K>, V> fetch(K from, K to, long timeFrom, long timeTo) {
```
nit: we generally make all params `final`

I guess there are other places, too

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java:
```diff
@@ -43,45 +43,58 @@ public CompositeReadOnlyWindowStore(final StateStoreProvider provider,
+        IteratorType empty();
+    }
+
+    public <ITERATORTYPE extends KeyValueIterator<?, V>> ITERATORTYPE fetch(Fetcher<K, V, ITERATORTYPE> fetcher) {
```
uggh - i think i saw you mentioned something about the checkstyle violation todo with generic types. Maybe we should change that rule? All caps is pretty ugly and Hard to read.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/FilteredCacheIterator.java:
```diff
@@ -24,11 +24,48 @@
+                          final CacheFunction cacheFunction) {
         this.cacheIterator = cacheIterator;
         this.hasNextCondition = hasNextCondition;
+        this.wrappedIterator = new PeekingKeyValueIterator<Bytes, LRUCacheEntry>() {
```
Could we not pass this iterator in via the ctor? Wouldn't need the `CacheFunction` anymore and would mean that we could write tests just for this if need be

xvrl replied:
I'd rather not, since the caller otherwise would have to know that both of those iterators need to share the same underlying iterator. I'd rather have this class encapsulate that.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/KeyValueIterators.java:
```diff
@@ -0,0 +1,71 @@
+
+import java.util.NoSuchElementException;
+
+public class KeyValueIterators {
```
nit: can be package-private

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreIterator.java:
```diff
@@ -22,17 +22,22 @@
-class MergedSortedCacheWindowStoreIterator<V> extends AbstractMergedSortedCacheStoreIterator<Long, Long, V> implements WindowStoreIterator<V> {
+class MergedSortedCacheWindowStoreIterator<V> extends AbstractMergedSortedCacheStoreIterator<Long, Long, V, byte[]> implements WindowStoreIterator<V> {
+
+    private final StateSerdes<Long, V> serdes;
```
All the subclasses of `AbstractMergedSortedCacheStoreIterator` need `serdes` any reason why you've pulled the field into the subclasses? Something to do with generics?

xvrl replied:
AbstractMergedSortedCacheStoreIterator doesn't need the serdes, since it delegates all the serialization / deserialization through its abstract methods. The main motivation was this bug https://github.com/apache/kafka/pull/3016 which was caused by calling serdes directly within the abstract class.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheWindowStoreKeyValueIterator.java:
```diff
@@ -0,0 +1,75 @@
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.StateSerdes;
+
+class MergedSortedCacheWindowStoreKeyValueIterator<K, V>
```
Can you add a unit test for this class?

xvrl replied:
done

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:
```diff
@@ -0,0 +1,76 @@
+
+import java.nio.ByteBuffer;
+
+public class SegmentedCacheFunction implements CacheFunction {
```
can we add a unit test for this class?

dguy replied:
does it need to be public?

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:
```diff
@@ -0,0 +1,76 @@
+    private final KeySchema keySchema;
+    private final long segmentInterval;
+
+    public SegmentedCacheFunction(KeySchema keySchema, long segmentInterval) {
```
nit: package-private?

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:
```diff
@@ -0,0 +1,76 @@
+        return keySchema.segmentTimestamp(key) / segmentInterval;
+    }
+
+    public int compareSegmentedKeys(Bytes cacheKey, Bytes storeKey) {
```
nit: package-private?

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java:
```diff
@@ -33,15 +39,59 @@ public void init(final String topic) {
     @Override
     public Bytes upperRange(final Bytes key, final long to) {
-        return WindowStoreUtils.toBinaryKey(key, to, Integer.MAX_VALUE, serdes);
+        final byte[] bytes = key.get();
```
is this and `lowerRange` identical to those in `SessionKeySchema`? If so then maybe we should extract them to avoid code duplication?

xvrl replied:
maxSuffix is different

dguy replied:
could still probably extract a method that removes almost all of the duplication?

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+
+    final KeyValueIterator<Bytes, byte[]> bytesIterator;
```
nit: `private` ?

xvrl replied:
I can make it protected, it's being used in the overridden methods above, I think IntelliJ wasn't smart enough to figure that out.

dguy replied:
yeah - probably

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreIteratorWrapper.java:
```diff
@@ -0,0 +1,195 @@
+
+import java.util.NoSuchElementException;
+
+class WindowStoreIteratorWrapper<K, V> {
```
would be good to add some unit tests for this

xvrl replied:
this is not actually a new class, it used to be `WrappedWindowStoreIterator`. I just renamed it since it's technically not an Iterator anymore, but instead provides wrapped iterators.

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java:
```diff
@@ -86,14 +90,26 @@ public void shouldPutFetchFromCache() throws Exception {
         assertFalse(b.hasNext());
-        assertEquals(3, cache.size());
+
+        final KeyValueIterator<Windowed<String>, Long> all = cachingStore.findSessions("a", "b", 0, 0);
```
can we extract these out into separate tests? The second and 3rd "tests" are testing a different function. I'd prefer this was three tests.

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java:
```diff
@@ -162,12 +178,18 @@ public void shouldIterateAcrossWindows() throws Exception {
         assertEquals(KeyValue.pair(DEFAULT_TIMESTAMP + WINDOW_SIZE, "b"), fetch.next());
         assertFalse(fetch.hasNext());
+
+        final KeyValueIterator<Windowed<String>, String> fetchRange =
```
Create another test for this?

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java:
```diff
@@ -201,6 +229,25 @@ public void shouldFetchAndIterateOverExactKeys() throws Exception {
         final List<KeyValue<Long, String>> expected = Utils.mkList(KeyValue.pair(0L, "0001"), KeyValue.pair(1L, "0003"), KeyValue.pair(60000L, "0005"));
         assertThat(toList(cachingStore.fetch("a", 0, Long.MAX_VALUE)), equalTo(expected));
+
+        assertThat(
```
Same here. Would prefer to see at least 2 tests, one for each of the `fetch` methods

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/SessionKeySchemaTest.java:
```diff
@@ -63,6 +63,25 @@ public void shouldFetchExactKeySkippingShorterKeys() throws Exception {
+    @Test
+    public void boundsSanity() throws Exception {
+        // easy
+        Bytes upper = sessionKeySchema.upperRange(Bytes.wrap(new byte[]{1}), Long.MAX_VALUE);
```
would prefer to see separate tests for each of these. Describing what they are testing

xvrl replied:
will do

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/WindowKeySchemaTest.java:
```diff
@@ -0,0 +1,61 @@
+    private final WindowKeySchema windowKeySchema = new WindowKeySchema();
+
+    @Test
+    public void boundsSanity() throws Exception {
```
as above

dguy reviewed (commented):
Thanks @xvrl left some more comments

xvrl committed:
saner and more consistent type parameter conventions

xvrl committed:
rely only on one hasNextCondition method

xvrl committed:
code nits

xvrl committed:
more explicit tests for window/session key ranges

xvrl committed:
add segmented cache function tests

xvrl committed:
add test for MergedSortedCacheWindowStoreKeyValueIterator

xvrl commented:
@dguy I believe I addressed all your comments

dguy reviewed (approved):
Thanks @xvrl - LGTM. @guozhangwang, @mjsax, @enothereska, @bbejeck can we please get a second review?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedCacheFunction.java:
```diff
@@ -0,0 +1,76 @@
+
+    static byte[] bytesFromCacheKey(Bytes cacheKey) {
+        byte[] binaryKey = new byte[cacheKey.get().length - SEGMENT_ID_BYTES];
+        System.arraycopy(cacheKey.get(), SEGMENT_ID_BYTES, binaryKey, 0, binaryKey.length);
```
We may need to quantify how much perf impact it may incur for such allocation calls (note we need to trigger once per cache put): this is not a comment for this PR specifically.

xvrl replied:
agree, when I was implementing this, I was thinking we should switch our serde to ByteBuffers, so we could more easily work with array slices, since we often just pass around portions of the underlying array when working with composite keys such as windowed / segmented keys.

guozhangwang replied:
Agreed, could you file a separate JIRA for this?

xvrl replied:
done https://issues.apache.org/jira/browse/KAFKA-5966

guozhangwang reviewed (approved):
LGTM! Big thanks to @xvrl for this huge contribution!

guozhangwang commented:
Merged to trunk.
